### PR TITLE
fqdn: Add Protocol to DNS Proxy Cache

### PR DESCRIPTION
**Pull request by nathanjsweet**
DNS Proxy indexes domain selectors by port
only. In cases where protocols collide on port
the DNS proxy may have a more restrictive selector than it should because it does not merge port
protocols for L7 policies (only ports).

Refactor all users of the DNS Proxy are updated
to add protocol to any DNS Proxy entries, and all
tests are updated to test for port-protocol
merge errors.

```release-note
fqdn: Fixed bug that caused DNS Proxy to be overly restrictive on allowed DNS selectors.
```

### CODE REVIEW

File: pkg/endpoint/policy.go
Comment (pippolo84) [inline pkg/endpoint/policy.go]: Nit: docstring should be updated to mention the returned protocol

File: pkg/fqdn/dnsproxy/helpers.go
Comment (pippolo84) [inline pkg/fqdn/dnsproxy/helpers.go]: Same here regarding the comment

File: pkg/fqdn/dnsproxy/helpers.go
Comment (pippolo84) [inline pkg/fqdn/dnsproxy/helpers.go]: `17` seems kind of a magic number, I think it'd be better to use `uint16(u8proto.UDP)` instead.

File: pkg/fqdn/dnsproxy/helpers.go
Comment (pippolo84) [inline pkg/fqdn/dnsproxy/helpers.go]: Same here: `6` -> `uint16(u8proto.TCP)`

File: pkg/fqdn/dnsproxy/helpers_test.go
Comment (pippolo84) [inline pkg/fqdn/dnsproxy/helpers_test.go]: Any particular reason to redefine these here instead of using the constants from `u8proto`? If we want to avoid repeating the type conversion at each call we might as well do:

```
const (
	tcpProto = uint16(u8proto.TCP)
	udpProto = uint16(u8proto.UDP)
)
```

File: pkg/fqdn/dnsproxy/proxy.go
Comment (pippolo84) [inline pkg/fqdn/dnsproxy/proxy.go:169]: comment above should be updated to mention the protocol, as well as the one for `portProtoToSelectorAllow` below

File: pkg/fqdn/dnsproxy/proxy.go
Comment (pippolo84) [inline pkg/fqdn/dnsproxy/proxy.go]: Nit: you can also do

```
ipRules, exists := p.restored[endpointID][destProto][destPort]
if !exists {
  return false
}
```

directly, since reading from a nil map is safe.

File: pkg/fqdn/dnsproxy/proxy.go
Comment (pippolo84) [inline pkg/fqdn/dnsproxy/proxy.go]: Just a note that this method is becoming increasingly complex to read and understand.
We should consider a bit of refactoring here and in `RestoreRules` below too, but that's material for a follow-up PR.

File: pkg/fqdn/dnsproxy/proxy.go
Comment (pippolo84) [inline pkg/fqdn/dnsproxy/proxy.go]: Shouldn't we `make` the inner map here before ranging over `portMap`?

`restoredRules[proto] = make(map[uint16][]restoredIPRule, len(portMap))`

Otherwise the write at line 350 will panic.

File: pkg/fqdn/dnsproxy/proxy.go
Comment (pippolo84) [inline pkg/fqdn/dnsproxy/proxy.go]: This can be simplified to:

```
	rules, exists = allow[endpointID][destProto][destPort]
	return
```

File: pkg/fqdn/restore/restore.go
Comment (pippolo84) [inline pkg/fqdn/restore/restore.go]: Not directly related to this PR, but if this should be used only for testing, we should consider adding a `*testing.T` parameter to its signature, in order to make it explicit that this is intended only for testing purposes (same for `IPRules.Sort`).

**Review by pippolo84 — CHANGES_REQUESTED** [review]
Thanks for the PR!

Note: the precheck error is due to a missing update of `MockFQDNProxy.UpdateAllowed` signature in `pkg/fqdn/proxy/proxy.go`.

**Comment by squeed** [comment]
Does it make sense to have a generic ProtoPort struct somewhere? That would make the signatures a lot simpler, as well as being a proper map key. It would also help prevent future errors where we talk about port without protocol.

**Comment by nathanjsweet** [comment]
This affects upgrades in a way that needs to be better fleshed out :cry:.

File: pkg/fqdn/restore/restore.go
Comment (christarazi) [inline pkg/fqdn/restore/restore.go:33]: Clever!

**Review by christarazi — COMMENTED** [review]
The backwards-compat restoration logic LGTM assuming manual validation. I only studied that given I was only focused on ensuring that upgrades are not broken in this area of the code (see https://github.com/cilium/cilium/pull/12964). I did not review the rest of the PR, so I will leave that to the existing assigned reviewers for that.

Also, it's worth mentioning that in future versions, we will not need the "migration" / version 1 code anymore and so we can remove it. Worth filing an issue so that we don't forget.

**Review by squeed — APPROVED** [review]

File: pkg/fqdn/restore/restore.go
Comment (nathanjsweet) [reply pkg/fqdn/restore/restore.go]: Done.

**Review by pippolo84 — APPROVED** [review]
Thanks! :100:

**Comment by nathanjsweet** [comment]
CI was in a bad state. Closing and reopening.

**Comment by nathanjsweet** [comment]
Quay was down.

**Pull request merged**
